File: backend/app/models/indent.py
import enum
from datetime import date, datetime
from decimal import Decimal
from typing import Optional

from sqlalchemy import DateTime, Date, Enum, ForeignKey, Index, Integer, Numeric, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import Base, TimestampMixin
from app.models.distributor_outlet import DistributorOutlet  # noqa: F401 (relationship target)
from app.models.user import User  # noqa: F401 (relationship target)


class IndentStatus(str, enum.Enum):
    PENDING = "pending"
    APPROVED = "approved"
    REJECTED = "rejected"


class Indent(Base, TimestampMixin):
    """One submitted indent from a Distributor Outlet: cylinders requested
    (filled) / returned (empty) per size, with the amount and payment status."""

    __tablename__ = "indents"

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    do_id: Mapped[int] = mapped_column(
        ForeignKey("distributor_outlets.id", ondelete="RESTRICT"),
        nullable=False, index=True,
    )
    indent_date: Mapped[date] = mapped_column(Date, nullable=False)

    total_stock: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    total_filled: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    total_empty: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    total_amount: Mapped[Decimal] = mapped_column(Numeric(12, 2), default=0, nullable=False)
    amount_paid: Mapped[Decimal] = mapped_column(Numeric(12, 2), default=0, nullable=False)
    balance: Mapped[Decimal] = mapped_column(Numeric(12, 2), default=0, nullable=False)

    created_by_id: Mapped[Optional[int]] = mapped_column(
        ForeignKey("users.id", ondelete="SET NULL"), nullable=True, index=True
    )

    status: Mapped[IndentStatus] = mapped_column(
        Enum(IndentStatus, name="indent_status", native_enum=False, length=16),
        default=IndentStatus.PENDING,
        nullable=False,
        index=True,
    )
    reviewed_by_id: Mapped[Optional[int]] = mapped_column(
        ForeignKey("users.id", ondelete="SET NULL"), nullable=True
    )
    reviewed_at: Mapped[Optional[datetime]] = mapped_column(DateTime(timezone=True), nullable=True)
    review_note: Mapped[Optional[str]] = mapped_column(Text, nullable=True)

    distributor_outlet: Mapped["DistributorOutlet"] = relationship(
        "DistributorOutlet", lazy="joined"
    )
    reviewer: Mapped[Optional["User"]] = relationship(
        "User", foreign_keys=[reviewed_by_id], lazy="joined"
    )
    items: Mapped[list["IndentItem"]] = relationship(
        back_populates="indent",
        cascade="all, delete-orphan",
        order_by="IndentItem.size_kg",
        lazy="selectin",
    )

    __table_args__ = (
        Index("ix_indents_do_date", "do_id", "indent_date"),
    )

    @property
    def do_code(self) -> str:
        return self.distributor_outlet.code

    @property
    def do_name(self) -> str:
        return self.distributor_outlet.owner_name

    @property
    def reviewed_by_name(self) -> Optional[str]:
        return self.reviewer.full_name if self.reviewer else None


class IndentItem(Base):
    __tablename__ = "indent_items"

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    indent_id: Mapped[int] = mapped_column(
        ForeignKey("indents.id", ondelete="CASCADE"), nullable=False, index=True
    )
    size_kg: Mapped[int] = mapped_column(Integer, nullable=False)
    stock: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    filled: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    empty: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    rate: Mapped[Decimal] = mapped_column(Numeric(12, 2), default=0, nullable=False)
    amount: Mapped[Decimal] = mapped_column(Numeric(12, 2), default=0, nullable=False)

    indent: Mapped[Indent] = relationship(back_populates="items")

File: app/lib/data/models/indent.dart
// Indent models — mirror `/api/indents`.

double _asDouble(dynamic v) {
  if (v == null) return 0.0;
  if (v is num) return v.toDouble();
  return double.tryParse('$v') ?? 0.0;
}

int _asInt(dynamic v) => v is num ? v.toInt() : int.tryParse('$v') ?? 0;

/// One cylinder size on the DO's indent form: how many it has sold (stock)
/// and the price per cylinder used to work out the amount.
class IndentSize {
  final int sizeKg;
  final int stock;
  final double rate;
  IndentSize({required this.sizeKg, required this.stock, required this.rate});

  factory IndentSize.fromJson(Map<String, dynamic> j) => IndentSize(
        sizeKg: _asInt(j['size_kg']),
        stock: _asInt(j['stock']),
        rate: _asDouble(j['rate']),
      );
}

class IndentItem {
  final int sizeKg;
  final int stock;
  final int filled;
  final int empty;
  final double rate;
  final double amount;
  IndentItem({
    required this.sizeKg,
    required this.stock,
    required this.filled,
    required this.empty,
    required this.rate,
    required this.amount,
  });

  factory IndentItem.fromJson(Map<String, dynamic> j) => IndentItem(
        sizeKg: _asInt(j['size_kg']),
        stock: _asInt(j['stock']),
        filled: _asInt(j['filled']),
        empty: _asInt(j['empty']),
        rate: _asDouble(j['rate']),
        amount: _asDouble(j['amount']),
      );
}

class Indent {
  final int id;
  final int doId;
  final String doCode;
  final String doName;
  final DateTime indentDate;
  final int totalStock;
  final int totalFilled;
  final int totalEmpty;
  final double totalAmount;
  final double amountPaid;
  final double balance;
  final String status; // pending | approved | rejected
  final String? reviewedByName;
  final DateTime? reviewedAt;
  final String? reviewNote;
  final List<IndentItem> items;

  Indent({
    required this.id,
    required this.doId,
    required this.doCode,
    required this.doName,
    required this.indentDate,
    required this.totalStock,
    required this.totalFilled,
    required this.totalEmpty,
    required this.totalAmount,
    required this.amountPaid,
    required this.balance,
    required this.status,
    this.reviewedByName,
    this.reviewedAt,
    this.reviewNote,
    required this.items,
  });

  bool get isPaid => balance <= 0;
  bool get isPending => status == 'pending';
  bool get isApproved => status == 'approved';
  bool get isRejected => status == 'rejected';

  factory Indent.fromJson(Map<String, dynamic> j) => Indent(
        id: _asInt(j['id']),
        doId: _asInt(j['do_id']),
        doCode: j['do_code'] as String? ?? '',
        doName: j['do_name'] as String? ?? '',
        indentDate: DateTime.parse(j['indent_date'] as String),
        totalStock: _asInt(j['total_stock']),
        totalFilled: _asInt(j['total_filled']),
        totalEmpty: _asInt(j['total_empty']),
        totalAmount: _asDouble(j['total_amount']),
        amountPaid: _asDouble(j['amount_paid']),
        balance: _asDouble(j['balance']),
        status: j['status'] as String? ?? 'pending',
        reviewedByName: j['reviewed_by_name'] as String?,
        reviewedAt: j['reviewed_at'] == null
            ? null
            : DateTime.tryParse(j['reviewed_at'] as String),
        reviewNote: j['review_note'] as String?,
        items: (j['items'] as List? ?? [])
            .map((e) => IndentItem.fromJson(Map<String, dynamic>.from(e as Map)))
            .toList(),
      );
}
